src/puppeteer_live_view_server.js

```diff
@@ -0,0 +1,467 @@
+     *
+     * Client side, the function is invoked by clicking the "back" button
+     * on a page detail (screenshot + HTML).
+     *
+     * @param {Page} page
+     */
+    stopCapturing(page) {
+        const id = this.pageIds.get(page);
+        this.removeAllListeners(id);
+    }
+
+    /**
+     * The getScreenshotAndHtml method simply retrieves the page's HTML
+     * content, takes a screenshot and returns both as a promise.
+     * Unfortunately, nothing prevents the Page from being closed while
+     * the screenshot is being taken, which results into error.
+     * Therefore, the method prevents the page from being closed
+     * by replacing its close method and handling the page close
+     * itself once the screenshot has been taken.
+     *
+     * @param {Page} page Puppeteer's Page
+     * @returns {Promise<Buffer>} screenshot
+     * @private
+     * @ignore
+     */
+    _getScreenshotAndHtml(page) {
+        // replace page's close function to prevent a close
+        // while the screenshot is being taken
+        const { close } = page;
+        let closed;
+        let closeArgs;
+        let closeResolve;
+        page.close = (...args) => {
+            if (!closed) closeArgs = args;
+            closed = true;
+            return new Promise((resolve) => {
+                closeResolve = resolve;
+            });
+        };
+
+        const cleanup = () => {
+            // replace the stolen close() method or call it,
+            // if it should've been called externally
+            if (closed) {
+                close.apply(page, closeArgs)
+                    .then(closeResolve);
+            } else {
+                page.close = close.bind(page);
+            }
+        };
+
+        return Promise.props({
+            image: page.screenshot(),
+            html: page.content(),
+        })
+            .timeout(this.screenshotTimeout, 'Puppeteer Live View: Screenshot timed out.')
+            .finally(cleanup);
+    }
+
+    /**
+     * Handler invoked whenever a Target is created in a Puppeteer's Browser.
+     * @param {Target} target
+     * @private
+     */
+    _onTargetCreated(target) {
+        if (target.type() === 'page') {
+            target.page()
+                .then((page) => {
+                    const id = `${this.id}_PAGE_${++this.pageIdCounter}`;
+                    this.pages.set(id, page);
+                    this.pageIds.set(page, id);
+                    this.emit('pagecreated', {
+                        id,
+                        browserId: this.id,
+                        url: page.url(),
+                    });
+                    page.on('load', () => {
+                        this.loadedPages.add(page); // page is loaded
+                    });
+                    // using "framenavigated" on main Frame since "targetchanged" does not seem reliable
+                    page.on('framenavigated', (frame) => {
+                        if (frame === page.mainFrame()) {
```

Comment:
I'd also return early here since there we are already at depth 5 in the code

Reply:
done